### ISSUE TITLE
stack buffer overflow in the JSON parser

### ISSUE THREAD

**Issue by lkoznjak** [CESNET/libyang#1039]
Hello,

it seems that a stack buffer overflow can occur in the JSON data parser. More specifically, in a call to **copyutf8** from **lyjson_parse_text**. Worth noting is that, when compiled without an address sanitizer, the program terminates normally and it doesn't seem to report the stack smashing (probably not reaching the canary). However, when compiled with ASAN, it reports a stack buffer overflow of one byte, which lands on the first byte of the stack canary. 

This is the AddressSanitizer output: 
```
=================================================================
==1281965==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd03b74070 at pc 0x7f3883101ab4 bp 0x7ffd03b73bb0 sp 0x7ffd03b73ba0
WRITE of size 1 at 0x7ffd03b74070 thread T0
    #0 0x7f3883101ab3 in copyutf8 /home/laki/libyang/src/parser.c:3821
    #1 0x7f3883154295 in lyjson_parse_text /home/laki/libyang/src/parser_json.c:160
    #2 0x7f38831590f4 in json_parse_data /home/laki/libyang/src/parser_json.c:879
    #3 0x7f388315ec65 in lyd_parse_json /home/laki/libyang/src/parser_json.c:1480
    #4 0x7f3883207813 in lyd_parse_ /home/laki/libyang/src/tree_data.c:967
    #5 0x7f3883207f3a in lyd_parse_data_ /home/laki/libyang/src/tree_data.c:1039
    #6 0x7f38832082f2 in lyd_parse_fd_ /home/laki/libyang/src/tree_data.c:1076
    #7 0x7f38832087d7 in lyd_parse_path /home/laki/libyang/src/tree_data.c:1119
    #8 0x55b5c5a811e6 in main (/home/laki/Documents/Sartura/fuzz/no-afl+0x11e6)
    #9 0x7f3882e87022 in __libc_start_main (/usr/lib/libc.so.6+0x27022)
    #10 0x55b5c5a8109d in _start (/home/laki/Documents/Sartura/fuzz/no-afl+0x109d)

Address 0x7ffd03b74070 is located in stack of thread T0 at offset 1056 in frame
    #0 0x7f388315389f in lyjson_parse_text /home/laki/libyang/src/parser_json.c:61

  This frame has 1 object(s):
    [32, 1056) 'buf' (line 64) <== Memory access at offset 1056 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/laki/libyang/src/parser.c:3821 in copyutf8
Shadow bytes around the buggy address:
  0x1000207667b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000207667c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000207667d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000207667e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000207667f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100020766800: 00 00 00 00 00 00 00 00 00 00 00 00 00 00[f3]f3
  0x100020766810: f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 00 00
  0x100020766820: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100020766830: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x100020766840: f1 f1 01 f2 04 f2 04 f2 00 f2 f2 f2 00 f2 f2 f2
  0x100020766850: 00 f2 f2 f2 00 f2 f2 f2 00 f3 f3 f3 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1281965==ABORTING
```

The project is compiled with an extra `-fsanitize=address` flag in the **CMakeLists.txt**, and the program is being executed with:
`LD_PRELOAD=/usr/lib/libasan.so ./overflow overflow-input.txt`

While the following code is being used to test the JSON data parser's behavior:
```
#include <stdio.h>
#include <libyang/libyang.h>

int main(int argc, char **argv) {
    struct ly_ctx *ctx = NULL;

    ctx = ly_ctx_new(NULL, 0);
    if (!ctx) {
        fprintf(stderr, "failed to create context.\n");
        return -1;
    }

    lyd_parse_path(ctx, argv[1], LYD_JSON, LYD_OPT_CONFIG);

    ly_ctx_destroy(ctx, NULL);

    return 0;
}
```

Attached is a file causing the stack buffer overflow.
[overflow-input.txt](https://github.com/CESNET/libyang/files/4365386/overflow-input.txt)

Regards,
Luka

**Comment by michalvasko** [comment]
Hi Luka,
okay, should be fixed.

Regards,
Michal

**Comment by lkoznjak** [comment]
A bit late for the reply, but that seems to do it, thank you!
Fixed by 369a4a8, closing the issue.
